macros/utils/cross_db_utils/get_profile_creation_query.sql

```diff
@@ -4,50 +4,80 @@
-{% for database in databases -%}
-GRANT USAGE,MONITOR ON DATABASE {{ database }} TO ROLE {{ parameters["role"] }};
-{%- endfor %}
-{% for schema_tuple in schema_tuples -%}
-GRANT USAGE,MONITOR ON SCHEMA {{ schema_tuple[0] }}.{{ schema_tuple[1] }} TO ROLE {{ parameters["role"] }};
-{%- endfor %}
+-- Set credentials as variables
+SET elementary_database = '{{ parameters["database"] }}';
+SET elementary_schema = '{{ parameters["schema"] }}';
+SET elementary_warehouse = '{{ parameters["warehouse"] }}';
+SET elementary_role = '{{ parameters["role"] }}';
+SET elementary_username = '{{ parameters["user"] }}';
+SET elementary_password = '{{ parameters["password"] }}';
+
+-- Account admin role required to set up permissions below
+USE ROLE ACCOUNTADMIN;
+
+-- Create elementary user and role
+CREATE OR REPLACE USER IDENTIFIER($elementary_username) PASSWORD = $elementary_password;
+CREATE OR REPLACE ROLE IDENTIFIER($elementary_role);
+GRANT ROLE IDENTIFIER($elementary_role) TO USER IDENTIFIER($elementary_username);
+
+-- Grant elementary role access to the supplied warehouse
+GRANT USAGE ON WAREHOUSE IDENTIFIER($elementary_warehouse) TO ROLE IDENTIFIER($elementary_role);
 
 -- Read access to elementary schema
-GRANT SELECT ON ALL TABLES IN SCHEMA {{ parameters["database"] }}.{{ parameters["schema"] }} TO ROLE {{ parameters["role"] }};
-GRANT SELECT ON FUTURE TABLES IN SCHEMA {{ parameters["database"] }}.{{ parameters["schema"] }} TO ROLE {{ parameters["role"] }};
-GRANT SELECT ON ALL VIEWS IN SCHEMA {{ parameters["database"] }}.{{ parameters["schema"] }} TO ROLE {{ parameters["role"] }};
-GRANT SELECT ON FUTURE VIEWS IN SCHEMA {{ parameters["database"] }}.{{ parameters["schema"] }} TO ROLE {{ parameters["role"] }};
-
--- Metadata access to rest of the schemas used within the dbt project
-{% for database, schema in schema_tuples -%}
-GRANT REFERENCES ON ALL TABLES IN SCHEMA {{ database }}.{{ schema }} TO ROLE {{ parameters["role"] }};
-GRANT REFERENCES ON FUTURE TABLES IN SCHEMA {{ database }}.{{ schema }} TO ROLE {{ parameters["role"] }};
-GRANT REFERENCES ON ALL VIEWS IN SCHEMA {{ database }}.{{ schema }} TO ROLE {{ parameters["role"] }};
-GRANT REFERENCES ON FUTURE VIEWS IN SCHEMA {{ database }}.{{ schema }} TO ROLE {{ parameters["role"] }};
+SET elementary_schema_fqn = $elementary_database || '.' || $elementary_schema;
+GRANT SELECT ON ALL TABLES IN SCHEMA IDENTIFIER($elementary_schema_fqn) TO ROLE IDENTIFIER($elementary_role);
+GRANT SELECT ON FUTURE TABLES IN SCHEMA IDENTIFIER($elementary_schema_fqn) TO ROLE IDENTIFIER($elementary_role);
+GRANT SELECT ON ALL VIEWS IN SCHEMA IDENTIFIER($elementary_schema_fqn) TO ROLE IDENTIFIER($elementary_role);
+GRANT SELECT ON FUTURE VIEWS IN SCHEMA IDENTIFIER($elementary_schema_fqn) TO ROLE IDENTIFIER($elementary_role);
+
+-- Information schema access
+CREATE OR REPLACE PROCEDURE ELEMENTARY_GRANT_INFO_SCHEMA_ACCESS(database_name STRING, role_name STRING)
```

Comment:
Changed this function to work by database instead of by schema because it should work as well.

The only edge case is actually with `FUTURE GRANTS` and the handling of it needs to be different (in Snowflake, when there are both database and schema future grants, all database future grants are being ignored).
It's a bit complicated to handle and not sure it's important enough for now so I omitted it.